src/map.rs

```diff
@@ -477,6 +477,39 @@ where
         }
     }
 
+    pub fn get_many_mut<'a, 'b, Q: ?Sized, const N: usize>(
+        &'a mut self,
+        keys: [&'b Q; N],
+    ) -> Option<[&'a mut V; N]>
+    where
+        Q: Hash + Equivalent<K>,
+    {
+        let indices = keys.map(|key| self.get_index_of(key));
+        if indices.iter().any(Option::is_none) {
+            return None;
+        }
+        let indices = indices.map(Option::unwrap);
+
+        // SAFETY: Can't allow duplicate indices as we would return several mutable refs to the same data
+        for i in 0..N {
+            let idx = indices[i];
+            if indices[i + 1..N].contains(&idx) {
+                return None;
+            }
+        }
+
+        let entries = self.as_entries_mut();
+        let out = indices.map(|i| {
+            // SAFETY: OK to discard mutable borrow lifetime as each index is unique
+            #[allow(unsafe_code)]
+            unsafe {
+                &mut *(&mut entries[i].value as *mut V)
```

Comment:
By the way, I did run miri on the tests like:
```
# rustup +nightly component add miri
cargo clean
cargo +nightly miri test
```
but I know miri doesn't catch all cases of undefined behaviour.

Reply:
Actually, maybe I understand. The variable `entries` is `&mut [Bucket<K, V>]` so it is a mutable reference the whole slice but the array `out` also holds mutable references to individual elements, which alias with `entries`. Is this what you are referring to?

Reply:
Yeah, that's it. I'm a little surprised that miri doesn't see any problem, but I think it's because slice indexing is an intrinsic operation in MIR, so miri doesn't really see any use of the entire `entries`. If you force it to go through the `IndexMut` trait, `entries.index_mut(i)`, then it does complain:

```
test map::tests::many_mut_multi_success ... error: Undefined Behavior: trying to reborrow <390898> for SharedReadOnly permission at alloc159164[0xc], but that tag does not exist in the borrow stack for this location
    --> /home/jistone/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/cmp.rs:1531:27
     |
1531 |             PartialEq::eq(*self, *other)
     |                           ^^^^^
     |                           |
     |                           trying to reborrow <390898> for SharedReadOnly permission at alloc159164[0xc], but that tag does not exist in the borrow stack for this location
     |                           this error occurs as part of a reborrow at alloc159164[0xc..0x10]
     |
     = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
     = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <390898> was created by a retag at offsets [0xc..0x10]
    --> src/map.rs:502:19
     |
502  |           let out = indices.map(|i| {
     |  ___________________^
503  | |             // SAFETY: OK to discard mutable borrow lifetime as each index is unique
504  | |             #[allow(unsafe_code)]
505  | |             unsafe {
506  | |                 &mut *(&mut entries.index_mut(i).value as *mut V)
507  | |             }
508  | |         });
     | |__________^
help: <390898> was later invalidated at offsets [0x0..0x40]
    --> src/map.rs:506:29
     |
506  |                 &mut *(&mut entries.index_mut(i).value as *mut V)
     |                             ^^^^^^^^^^^^^^^^^^^^
     = note: backtrace: [...]
    --> src/map/tests.rs:449:5
     |
449  |     assert_eq!(map.get_many_mut([&1, &1123]), Some([&mut 10, &mut 100]));
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

> > We probably need a manual bounds-check too...
> 
> Yeah that makes sense if pointers are the way to go but it should still panic rather than return None, I assume?

Yeah, if `get_index_of` returns a bad index, then we have internal errors, so a panic is fine.